Add date shell commands

nixward opened the pull request:
Currently only supports Unix time.

'get' subcommand returns date in format “Y-m-d H:M:S”
'set' subcommand format is '[Y-m-d] <H:M:S>'

The 'set' subcommand is implemented with basic date validation.

For user convenience of small adjustments to time the time argument will
accept H:M:S, :M:S or ::S where the missing field(s) will be filled in by
the previous time state.

nixward committed:
shell: modules: Add date commands to display or set date

Currently only supports Unix time.

‘get’ subcommand returns date in format “Y-m-d H:M:S”
‘set’ subcommand format is ‘[Y-m-d] <H:M:S>’

The ‘set’ subcommand is implemented with basic date validation.

For user convenience of small adjustments to time the time argument
will accept H:M:S, :M:S or ::S where the missing field(s) will be
filled in by the previous time state.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

nixward committed:
samples: shell_module: Add date commands

Add date command to set and display date.

Signed-off-by: Nick Ward <nix.ward@gmail.com>

jakub-uC reviewed (commented):
Most likely I am missing something basic here but why you allow setting minutes and seconds in different ranges?

nixward commented:
> Most likely I am missing something basic here but why you allow setting minutes and seconds in different ranges?

I feel a little uneasy about this also.  I was just allowing for the very unlikely event someone wants to set it at the moment of a leap second.  Weighting it up I didn't think there would be any harm to allow it to be a bit more permissive with the small chance someone puts their clock out by a second.  It does contradict the fact this PR's current scope is to only support Unix time.  If you think this is bad I can definitely change this.  There is currently no concise validation of month lengths (which could possibly come in a separate PR) making the day field extra permissive as well.  I'll add comments to address this but I'm also happy to restrict it.

jakub-uC commented:
> > Most likely I am missing something basic here but why you allow setting minutes and seconds in different ranges?
> 
> I feel a little uneasy about this also. I was just allowing for the very unlikely event someone wants to set it at the moment of a leap second. Weighting it up I didn't think there would be any harm to allow it to be a bit more permissive with the small chance someone puts their clock out by a second. It does contradict the fact this PR's current scope is to only support Unix time. If you think this is bad I can definitely change this. There is currently no concise validation of month lengths (which could possibly come in a separate PR) making the day field extra permissive as well. I'll add comments to address this but I'm also happy to restrict it.

I will appreciate a comment in the code :)
`LGTM`

pabigot commented on subsys/shell/modules/date_service.c:
```diff
@@ -0,0 +1,117 @@
+	ARG_UNUSED(argc);
+
+	tm.tm_year = strtol(argv[1], NULL, 10);
+	if ((tm.tm_year < 1900) || (tm.tm_year > 9999)) {
```
This is too strict; years before 1900 are allowed, and there is no documented limit on an upper bound [that I can find](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/time.h.html).  Just accept any value that made it through `strtol`.

nixward replied:
Done

pabigot commented on subsys/shell/modules/date_service.c:
```diff
@@ -0,0 +1,117 @@
+{
+	shell_print(shell,
+		    "%d-%02d-%02d "
+		    "%02d:%02d:%02d UTC",
```
Except for year these should use the `u` conversion specifier rather than `d` because they can't legitimately be negative.

nixward replied:
Done

pabigot commented on subsys/shell/modules/date_service.c:
```diff
@@ -0,0 +1,117 @@
+		return -EINVAL;
+	}
+	tm.tm_sec = strtol(argv[6], NULL, 10);
+	/* Note range allows for a leap second */
```
Thank you for respecting leap seconds even though nobody's ever going to enter one by hand.

pabigot commented on subsys/shell/modules/date_service.c:
```diff
@@ -0,0 +1,117 @@
+
+	ARG_UNUSED(argc);
+
+	tm.tm_year = strtol(argv[1], NULL, 10);
```
I personally think it's worth validating the input:
```
char *ep = NULL;
val = strtol(argv[i], &ep, 10);
if (!(ep && ((*ep == 0) || (*ep == ':'))) {
  return -EINVAL;
}
ep = NULL;
```

Especially if you accept the proposal to make the user input more flexible (checking termination on `:` or `-` for non-final fields).

nixward replied:
I've kept the input format to match the output format (specific `-` and `:` to Y-m-d and H:M:S respectively and included validation as required.

pabigot reviewed (changes requested):
I haven't tested this personally but it looks like a good step towards improving civil time support.

I think it would be worth reconsidering the API for setting the clock to be more helpful.  The set format "Y m d H M S" is not consistent with the display format `YY-mm-dd HH:MM:SS UTC`.  Also consider the use case: it may be that most times people set the time they're trying to correct a slight error in an already-close time (e.g. from a battery-backed clock).

Could you accept something like `[Y-m-d] H:M:S` where the date part is completely optional, and any of the values (but not punctuation) that are missing are filled in by the current time?  Then if the system time is `2020-02-10 12:34:56` and the actual time is `2020-02-10 12:43:20` the user can update it with the string ":43:20".

The year validation and use of unsigned conversion specifiers should be changed.  The rework of user interface to be more flexible is suggestion, not a requirement; if you don't want it I'll approve anyway.

nixward commented:
@pabigot thanks for the review, I'll address what you have raised ASAP.  Would you have advice on how to get shippable to pass?  There's seems to be an issue with the timespec header for some builds.

pabigot commented:
> Would you have advice on how to get shippable to pass? There's seems to be an issue with the timespec header for some builds.

Yeah.  I don't have a solution for that; AFAICT `<sys/timespec.h>` is a BSD header, not a POSIX header, so I don't know why the Zephyr POSIX subsystem is including it for native builds.  There's some discussion in 0173d86e6fc09adf76a151e54385f968dc79a79d.

#18892 may also be relevant, as it would eliminate the need for Zephyr's `<time.h>` to try to reconstruct the content of a system `<time.h>`

pfalcon commented:
> Yeah. I don't have a solution for that; AFAICT <sys/timespec.h> is a BSD header, not a POSIX header, so I don't know why the Zephyr POSIX subsystem

Probably because "Zephyr POSIX subsystem" is intended to be real-worldly useful, e.g. interoperate with other components (e.g. newlib), and be able to build real-world software?

> is including it for native builds. 

"native builds" mentioned here are actually non-native builds, from Zephyr PoV. So-called "native_posix" target is a very special, use-only-if-you-know-what-you're-doing target which tries to emulate a **subset** of Zephyr functionality on top of a foreign system (full-fledged POSIX in this cass).  It's not compatible with various Zephyr subsystems. For one, with Zephyr POSIX subsystem.

nixward commented:
@pfalcon would you have advice on how to get the shippable build failure to pass?

pfalcon commented:
@pabigot: If the date handling functionality implemented here is ok by you, can you approve or at least dismiss your -1?

pfalcon commented:
> @pfalcon would you have advice on how to get the shippable build failure to pass?

There're generally 2 ways:

1. Blacklist CONFIG_ARCH_POSIX-based systems (as they're not compatible with Zephyr POSIX subsys).
2. Given that here only clock_gettime/clock_settime is used, try to workaround ARCH_POSIX handling of these calls.

p.2 definitely in scope, there's anticipation and provisions that various software may require just small bits of POSIX, instead of a full coherent subsys, so there're ways to enable individual pieces (at the full responsibility and risk of particular target applications, there's no guarantee it would work in general case).

But "only clock_gettime/clock_settime" is deceptive, because handling happens at the granularity of headers, and the corresponding header(s) define much more stuff.

So, let's specify the mission statement (with special attn: to @pabigot, as you regularly question why sth in POSIX subsys was done in this or that way, please follow thru this case, in all previous cases, similar handling/logic was applied): currently Zephyr POSIX subsys supports 2 libc's: Newlin and minlibc. Here, we're talking about adding bits of support for 3rd libc - Linux' Glibc.

Let me remind how support for Newlib and minlibc was implemented: Newlib (as a more developed libc) was taken as a model, and minlibc was made to be consistent with it. In this case, Newlib defined `struct timespec` in `<sys/timespec.h>`. So, minlibc was made to do the same, then the same include covered both.

As can be seen from build error here, Linux doesn't have that header. So, let's now see how Linux Glibc-based system deals with that define. A grep here shows that it's defined in /usr/include/x86_64-linux-gnu/bits/types/struct_timespec.h. It's important to note that the pattern to resolve this issue is the same - each "system-specific" type is defined in a separate header file, or there would be untanglable mess. If only location for these system-specific headers was standardized! But nope, so where we use `<sys/timespec.h>` for newlib/minlibc, we'll need to #ifdef to use `<bits/types/struct_timespec.h>`. Oh, and of course, that header belongs not to Glibc, but to Linux kernel external API headers, so our mission statement is adjusted to "... interface with Linux kernel defines".

Repeated a bunch of times for other types, not used in this PR, but pulled from `<time.h>`, and thus leading to conflicts, I was able to build this PR for BOARD=native_posix. I'll put up my changes into a brunch shortly, and let's keep fingers crossed that fixing one thing didn't broke others (because that's how it typically was in POSIX subsys, until disciplined patterns like "minlibc should follow Newlib as close as possible" were adopted).

pfalcon commented on subsys/shell/modules/Kconfig:
```diff
@@ -13,3 +13,8 @@ config DEVICE_SHELL
 	  This shell provides access to basic device data.
+
+config DATE_SHELL
+	bool "Enable date shell"
```
There definitely should be `depends on POSIX_CLOCK` here.

nixward replied:
Thanks, done!

pabigot reviewed (approved):
Thanks for the changes; this looks good.

I was going to ask that `-ENOEXEC` be replaced by `-EINVAL` for all detected format/input errors, but AFAICT this mis-use is pervasive in Zephyr shell so for consistency I suppose it should stay.

pfalcon committed:
include: posix: time: Make possible to include with CONFIG_ARCH_POSIX

While CONFIG_ARCH_POSIX in general isn't compatible with Zephyr POSIX
subsys (because CONFIG_ARCH_POSIX itself is implemented on top of
POSIX, so there're obvious conflicts), apply workaround to allow to
at least use clock_gettime() and clock_settime() functions.

This change is grounded in upcoming support for date manipulation
commands for Zephyr shell, which are implemented using functions
above. There's no guarantee that CONFIG_ARCH_POSIX and Zephyr POSIX
subsys will coexist for any other usecase. (But the change is
relatively clean and is definitely in the right direction of
prototyping ways of such a coexistance.)

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>
(cherry picked from commit 4a05644bce81f4792c018964d29c1158f1982d1e)

pfalcon commented:
Ok, I submitted my changes as https://github.com/zephyrproject-rtos/zephyr/pull/23219 to check that it doesn't cause regression. Build passed. The suggested action is to pick that commit into that PR. The above PR is marked as DNM.

aescolar commented:
> "native builds" mentioned here are actually non-native builds, from Zephyr PoV. So-called "native_posix" target is a very special, use-only-if-you-know-what-you're-doing target which tries to emulate a subset of Zephyr functionality on top of a foreign system (full-fledged POSIX in this cass). It's not compatible with various Zephyr subsystems. For one, with Zephyr POSIX subsystem.

This is quite misleading: The native_posix port does not "emulate Zephyr". It implements a replacement arch layer (CPU emulation if you wish) and a few peripherals, so the Zephyr kernel runs on it unknowingly. The only subsystems it is not compatible with are the POSIX shim*1 and userspace*2, or those requiring drivers which are not yet implemented in native_posix (The full list is actually here #6044).
You can consider it a means that allows for integration tests and which has a few of the same practical limitation you would face with a unit test framework (as you are compiling your code targeting your workstation).

*1 which was supported for a while. But on @pfalcon 's request, to simplify the development of the POSIX API, support was dropped.
*2 there is no MPU emulation.

nixward commented:
> I was going to ask that `-ENOEXEC` be replaced by `-EINVAL` for all detected format/input errors, but AFAICT this mis-use is pervasive in Zephyr shell so for consistency I suppose it should stay.

I see what you are saying now.  I've switched it.  If any one objects let me know.

pfalcon commented:
@nixward : I guess DNM label should be removed now.
